### PR TITLE
Set discovery tool to close with SIGINT [9336]

### PR DESCRIPTION
**Pull request by jparisu**
Disccovery tool closed with a keyboard input. This broke with background run.
It has changed to close with SIGINT.
Test added to test this functionality.

### CODE REVIEW

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/2618/)
* Linux aarch64[Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux%20aarch64/39/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/2433/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/2628/)

**Comment by MiguelCompany** [comment]
The new test `test_fast_discovery_closure` fails on Windows and Mac

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/2830/)
* Linux aarch64[Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux%20aarch64/165/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/2596/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/2846/)

**Comment by MiguelCompany** [comment]
* New test still fails on Windows.
* There are some uncrustify issues

I've marked this as no-test, so we can manually run windows only

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/3047/)
* Linux aarch64Build Status
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/2731/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/3058/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/3058/)
* Linux aarch64Build Status
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/2740/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/3069/)

File: test/system/tools/fds/Testing/Temporary/CTestCostData.txt [hunk 1]
```diff
@@ -0,0 +1 @@
+---
```
Comment (IkerLuengo) [inline test/system/tools/fds/Testing/Temporary/CTestCostData.txt:1]: Testing directory should not be added to the repo. Maybe we need a rule in the .gitignore of FDS

File: test/system/tools/fds/tests.py [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+# limitations under the License.
+
+"""
+    Tests for the fastdds tool.
+
+    Contains a package of system test for fastdds tool
+
+    usage: test.py <binary_path> <test_name>
+
+    binary: Fast-DDS binary path
```
Comment (IkerLuengo) [inline test/system/tools/fds/tests.py:22]:
```suggestion
    binary_path: Fast-DDS binary path
```

File: test/system/tools/fds/tests.py [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+            # sleep to let the server shut down
+            time.sleep(1)
+
+    except KeyboardInterrupt:
+       print('Catched a KeyboardInterrupt')
+       pass
+
+    # CTest behaves differently than calling python directly
+
+    EXPECTED_CLOSURE = "### Server shutted down ###"
```
Comment (IkerLuengo) [inline test/system/tools/fds/tests.py:73]:
```suggestion
    EXPECTED_CLOSURE = "### Server shut down ###"
```

File: tools/fds/server.cpp [hunk 4]
```diff
@@ -111,11 +130,10 @@ int main (
...
 
-    fastdds::dds::Log::SetCategoryFilter(
-        std::regex("(RTPS_PDPSERVER_TRIM)|(RTPS_PARTICIPANT)|(DISCOVERY_SERVER)"
-        "|(SERVER_PDP_THREAD)|(CLIENT_PDP_THREAD)|(RTPS_WRITER_HISTORY)|(RTPS_READER_HISTORY)|(RTPS_PDP)"
-        "|(DISCOVERY_DATABASE)|(RTPS_PDP_LISTENER)|(SERVER_PING_THREAD)"));
-    fastdds::dds::Log::SetVerbosity(fastdds::dds::Log::Kind::Info);
+    //  fastdds::dds::Log::SetCategoryFilter(
+    //      std::regex("(RTPS_HISTORY)|(RTPS_WRITER_HISTORY)|(RTPS_READER_HISTORY)|(RTPS_PDP_SERVER)|(READER_PROXY)"
+    //      "|(RTPS_PDP)|(SERVER_PDP_THREAD)|(CLIENT_PDP_THREAD)|(DISCOVERY_DATABASE)|(RTPS_PDP_LISTENER)"));
+    //  fastdds::dds::Log::SetVerbosity(fastdds::dds::Log::Kind::Info);
```
Comment (IkerLuengo) [inline tools/fds/server.cpp:136]: Remove this block

File: tools/fds/server.cpp [hunk 5]
```diff
@@ -188,16 +207,20 @@ int main (
...
+        cout << "\n### Server is running ###" << std::endl;
+
+        g_signal_cv.wait(lock, []
+                {
+                    return 0 != g_signal_status;
+                });
 
-        // Remove the server
-        Domain::removeParticipant(pServer);
+        cout << "\n### Server shutted down ###" << std::endl;
```
Comment (IkerLuengo) [inline tools/fds/server.cpp:221]:
```suggestion
        cout << "\n### Server shut down ###" << std::endl;
```

File: tools/fds/server.cpp [hunk 6]
```diff
@@ -188,16 +207,20 @@ int main (
...
+        signal(SIGINT, sigint_handler);
+        cout << "\n### Server is running ###" << std::endl;
+
+        g_signal_cv.wait(lock, []
+                {
+                    return 0 != g_signal_status;
+                });
 
-        // Remove the server
-        Domain::removeParticipant(pServer);
```
Comment (IkerLuengo) [inline tools/fds/server.cpp:197]: Shouldn't we remove the Participant for a clean exit, anyways?

**Review by IkerLuengo — CHANGES_REQUESTED** [review]
I'm trusting @MiguelBarro with the PS scripting, his knowledge in that field is far better than mine.

File: test/system/tools/fds/Testing/Temporary/CTestCostData.txt (hunk 1 above)
Comment (MiguelBarro) [reply test/system/tools/fds/Testing/Temporary/CTestCostData.txt:1]: 👍
 `>git rm test\system\tools\fds\Testing -r`

File: test/system/tools/fds/tests.py (hunk 2 above)
Comment (MiguelBarro) [reply test/system/tools/fds/tests.py:22]: 👍

File: test/system/tools/fds/tests.py (hunk 3 above)
Comment (MiguelBarro) [reply test/system/tools/fds/tests.py:73]: That's a good one. It's modified also in server.cpp for correction and in order to keep the logic.

File: tools/fds/server.cpp (hunk 4 above)
Comment (MiguelBarro) [reply tools/fds/server.cpp:136]: Somehow I feel I'm gonna miss it the next bug hunt.🤠

File: tools/fds/server.cpp (hunk 5 above)
Comment (MiguelBarro) [reply tools/fds/server.cpp:221]: 👍

File: tools/fds/server.cpp (hunk 6 above)
Comment (MiguelBarro) [reply tools/fds/server.cpp:197]: The later called to `Domain::stopAll()` takes care of it.

File: test/system/tools/fds/CMakeLists.txt [hunk 7]
```diff
@@ -24,21 +24,66 @@ if(PYTHONINTERP_FOUND)
...
+
+        if(TARGET tinyxml2 OR TARGET tinyxml2::tinyxml2)
+            set(TEST_ENVIRONMENT 
+                "${TEST_ENVIRONMENT};$<TARGET_FILE_DIR:${TINYXML2_LIBRARY}>")
+        elseif(EXISTS TINYXML2_LIBRARY) 
+            get_filename_component(
+                TINYXML2_LIBRARY_DIR ${TINYXML2_LIBRARY} DIRECTORY)
+            set(TEST_ENVIRONMENT 
+                "${TEST_ENVIRONMENT};${TINYXML2_LIBRARY_DIR}")
+            unset(TINYXML2_LIBRARY_DIR)
```
Comment (jparisu) [inline test/system/tools/fds/CMakeLists.txt:46]: You may be unsetting an env variable that could be use somewhere else. Not sure what is the propouse, a comment will be helpful.

File: test/system/tools/fds/CMakeLists.txt [hunk 8]
```diff
@@ -24,21 +24,66 @@ if(PYTHONINTERP_FOUND)
...
+    # windows auxiliary script to fork test execution
+    set(PWS_LAUNCHER
+        ${CMAKE_CURRENT_SOURCE_DIR}/launcher.ps1
+    )
+
+    # Calculate environment
+    set(TEST_ENVIRONMENT 
+        "PATH=$ENV{PATH};$<TARGET_FILE_DIR:fastcdr>;$<TARGET_FILE_DIR:fastrtps>")
+
+    if(WIN32)
```
Comment (jparisu) [inline test/system/tools/fds/CMakeLists.txt:36]: You could slightly comment why this is necessary.

**Review by jparisu — COMMENTED** [review]
Some comments in `CMakeLists.txt` will save a lot of time for a future modification or update of the test.

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/3061/)
* Linux aarch64[Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux%20aarch64/282/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/2743/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/3072/)

File: test/system/tools/fds/CMakeLists.txt (hunk 7 above)
Comment (MiguelBarro) [reply test/system/tools/fds/CMakeLists.txt:46]: I only change the environment for this specific test.
TEST_ENVIRONMENT has directory scope cannot be propagated upwards, thus, no side effects.

File: test/system/tools/fds/CMakeLists.txt (hunk 8 above)
Comment (MiguelBarro) [reply test/system/tools/fds/CMakeLists.txt:36]: Everybody knows we must populate the environment to help a binary locate its dependencies. That's why we use local_setup.sh or setup.sh in colcon.

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/3066/)
* Linux aarch64[Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux%20aarch64/286/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/2747/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/3079/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/3068/)
* Linux aarch64[Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux%20aarch64/288/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/2749/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/3084/)

**Review by IkerLuengo — DISMISSED** [review]
LGTM

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/3072/)
* Linux aarch64[Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux%20aarch64/292/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/2752/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/3089/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/3081/)
* Linux aarch64[Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux%20aarch64/298/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/2757/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/3102/)

**Comment by MiguelCompany** [comment]
Failures are unrelated

**Comment by MiguelBarro** [comment]
> Failures are unrelated

FastRTPS Manual Windows job #3103 witness (running 100 times the test in each platform) that windows flackiness has disappear.

http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/3103/

**Review by IkerLuengo — APPROVED** [review]
LGTM

**Pull request merged**